data/transactions/verify/streamverifier.go

```diff
@@ -295,127 +195,40 @@ func (sv *StreamVerifier) tryAddVerificationTaskToThePool(ue []*UnverifiedElemen
 }
 
-func (sv *StreamVerifier) addVerificationTaskToThePoolNow(ue []*UnverifiedElement) error {
+func (sv *StreamVerifier) addVerificationTaskToThePoolNow(unvrifiedElts []UnverifiedElement) error {
 	// if the context is canceled when the task is in the queue, it should be canceled
 	// copy the ctx here so that when the StreamVerifier is started again, and a new context
 	// is created, this task still gets canceled due to the ctx at the time of this task
 	taskCtx := sv.ctx
 	function := func(arg interface{}) interface{} {
+		uElmts := arg.([]UnverifiedElement)
 		if taskCtx.Err() != nil {
 			// ctx is canceled. the results will be returned
-			sv.cleanup(ue)
+			sv.ep.Cleanup(uElmts, ErrShuttingDownError)
 			return nil
 		}
 
-		ue := arg.([]*UnverifiedElement)
-		batchVerifier := crypto.MakeBatchVerifier()
-
-		bl := makeBatchLoad(len(ue))
-		// TODO: separate operations here, and get the sig verification inside the LogicSig to the batch here
-		blockHeader := sv.nbw.getBlockHeader()
-		for _, ue := range ue {
-			groupCtx, err := txnGroupBatchPrep(ue.TxnGroup, blockHeader, sv.ledger, batchVerifier, nil)
-			if err != nil {
-				// verification failed, no need to add the sig to the batch, report the error
-				sv.sendResult(ue.TxnGroup, ue.BacklogMessage, err)
-				continue
-			}
-			totalBatchCount := batchVerifier.GetNumberOfEnqueuedSignatures()
-			bl.addLoad(ue.TxnGroup, groupCtx, ue.BacklogMessage, totalBatchCount)
-		}
+		batchVerifier, ctx := sv.ep.PreProcessUnverifiedElements(uElmts)
 
 		failed, err := batchVerifier.VerifyWithFeedback()
 		// this error can only be crypto.ErrBatchHasFailedSigs
-		if err == nil { // success, all signatures verified
-			for i := range bl.txnGroups {
-				sv.sendResult(bl.txnGroups[i], bl.elementBacklogMessage[i], nil)
-			}
-			sv.cache.AddPayset(bl.txnGroups, bl.groupCtxs)
-			return nil
-		}
 
-		verifiedTxnGroups := make([][]transactions.SignedTxn, 0, len(bl.txnGroups))
-		verifiedGroupCtxs := make([]*GroupContext, 0, len(bl.groupCtxs))
-		failedSigIdx := 0
-		for txgIdx := range bl.txnGroups {
-			txGroupSigFailed := false
-			for failedSigIdx < bl.messagesForTxn[txgIdx] {
-				if failed[failedSigIdx] {
-					// if there is a failed sig check, then no need to check the rest of the
-					// sigs for this txnGroup
-					failedSigIdx = bl.messagesForTxn[txgIdx]
-					txGroupSigFailed = true
-				} else {
-					// proceed to check the next sig belonging to this txnGroup
-					failedSigIdx++
-				}
-			}
-			var result error
-			if !txGroupSigFailed {
-				verifiedTxnGroups = append(verifiedTxnGroups, bl.txnGroups[txgIdx])
-				verifiedGroupCtxs = append(verifiedGroupCtxs, bl.groupCtxs[txgIdx])
-			} else {
-				result = err
-			}
-			sv.sendResult(bl.txnGroups[txgIdx], bl.elementBacklogMessage[txgIdx], result)
-		}
-		// loading them all at once by locking the cache once
-		sv.cache.AddPayset(verifiedTxnGroups, verifiedGroupCtxs)
+		sv.ep.PostProcessVerifiedElements(ctx, failed, err)
```

Comment:
Figuring out the failed signatures and pushing the results out a chan can be shared between votes and sigs.

However, for sigs, the verified sigs need to be cached, to avoid re-validation when validating the block. This caching is done with metadata obtained from the pre-processing (i.e. extracting sigs from the txnGroup). 

As a result: (1) the pre-processing phase cannot be generalized because of extracting sigs and metadata from txnGroups; (2) the post-processing phase cannot be generalized because of the caching.